common/rfb/SSecurityRSAAES.cxx

```diff
@@ -0,0 +1,587 @@
+    (rdr::U8)(len & 0xff)
+  };
+  int hashSize;
+  if (keySize == 128) {
+    hashSize = 20;
+    struct sha1_ctx ctx;
+    sha1_init(&ctx);
+    sha1_update(&ctx, 4, lenServerKey);
+    sha1_update(&ctx, serverKey.size, serverKeyN);
+    sha1_update(&ctx, serverKey.size, serverKeyE);
+    sha1_update(&ctx, 4, lenClientKey);
+    sha1_update(&ctx, clientKey.size, clientKeyN);
+    sha1_update(&ctx, clientKey.size, clientKeyE);
+    sha1_digest(&ctx, hashSize, hash);
+  } else {
+    hashSize = 32;
+    struct sha256_ctx ctx;
+    sha256_init(&ctx);
+    sha256_update(&ctx, 4, lenServerKey);
+    sha256_update(&ctx, serverKey.size, serverKeyN);
+    sha256_update(&ctx, serverKey.size, serverKeyE);
+    sha256_update(&ctx, 4, lenClientKey);
+    sha256_update(&ctx, clientKey.size, clientKeyN);
+    sha256_update(&ctx, clientKey.size, clientKeyE);
+    sha256_digest(&ctx, hashSize, hash);
+  }
+  raos->writeBytes(hash, hashSize);
+  raos->flush();
+}
+
+bool SSecurityRSAAES::readHash()
+{
+  rdr::U8 hash[32];
+  rdr::U8 realHash[32];
+  int hashSize = keySize == 128 ? 20 : 32;
+  if (!rais->hasData(hashSize))
+    return false;
+  rais->readBytes(hash, hashSize);
+  size_t len = serverKeyLength;
+  rdr::U8 lenServerKey[4] = {
+    (rdr::U8)((len & 0xff000000) >> 24),
+    (rdr::U8)((len & 0xff0000) >> 16),
+    (rdr::U8)((len & 0xff00) >> 8),
+    (rdr::U8)(len & 0xff)
+  };
+  len = clientKeyLength;
+  rdr::U8 lenClientKey[4] = {
+    (rdr::U8)((len & 0xff000000) >> 24),
+    (rdr::U8)((len & 0xff0000) >> 16),
+    (rdr::U8)((len & 0xff00) >> 8),
+    (rdr::U8)(len & 0xff)
+  };
+  if (keySize == 128) {
+    struct sha1_ctx ctx;
+    sha1_init(&ctx);
+    sha1_update(&ctx, 4, lenClientKey);
+    sha1_update(&ctx, clientKey.size, clientKeyN);
+    sha1_update(&ctx, clientKey.size, clientKeyE);
+    sha1_update(&ctx, 4, lenServerKey);
+    sha1_update(&ctx, serverKey.size, serverKeyN);
+    sha1_update(&ctx, serverKey.size, serverKeyE);
+    sha1_digest(&ctx, hashSize, realHash);
+  } else {
+    struct sha256_ctx ctx;
+    sha256_init(&ctx);
+    sha256_update(&ctx, 4, lenClientKey);
+    sha256_update(&ctx, clientKey.size, clientKeyN);
+    sha256_update(&ctx, clientKey.size, clientKeyE);
+    sha256_update(&ctx, 4, lenServerKey);
+    sha256_update(&ctx, serverKey.size, serverKeyN);
+    sha256_update(&ctx, serverKey.size, serverKeyE);
+    sha256_digest(&ctx, hashSize, realHash);
+  }
+  if (memcmp(hash, realHash, hashSize) != 0)
+    throw ConnFailedException("hash doesn't match");
+  return true;
+}
+
+void SSecurityRSAAES::writeSubtype()
+{
+  if (NeedUserName)
+    raos->writeU8(secTypeRA2UserPass);
+  else
+    raos->writeU8(secTypeRA2Pass);
+  raos->flush();
+}
+
+bool SSecurityRSAAES::readCredentials()
+{
+  if (!rais->hasData(1))
+    return false;
+  const rdr::U8* ptr = rais->getptr(1);
+  rdr::U8 lenUsername = ptr[0];
+  if (!rais->hasData(1 + lenUsername + 1))
+    return false;
+  ptr = rais->getptr(1 + lenUsername + 1);
```

Comment:
This is precisely the kind of thing that `hasDataOrRestore()` was designed to handle. So please use that so this code is consistent with other protocol code.